clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -1523,9 +1570,25 @@ void Sema::CheckCompleteDecompositionDeclaration(DecompositionDecl *DD) {
   // If the type of the decomposition is dependent, then so is the type of
   // each binding.
   if (DecompType->isDependentType()) {
-    for (auto *B : DD->bindings())
-      B->setType(Context.DependentTy);
+    for (auto *B : DD->bindings()) {
+      // Do not overwrite any pack type.
+      if (B->getType().isNull())
+        B->setType(Context.DependentTy);
+    }
```

Comment:
I really would like to understand when this happen. Maybe a longer comment is in order?

Reply:
This was here before. After processing the intializer the BindingDecls should all have types set except in the case when the initializer is dependent. In that case all the types are still Null so here we set them to dependent. My change is just preventing it from overwriting the PackExpansionType if there is one.

Reply:
The comment I updated is above this.